python/ray/_private/runtime_env/pip.py

```diff
@@ -50,36 +47,194 @@ def get_uri(runtime_env: Dict) -> Optional[str]:
+        # TODO(fyrestone): Support -i, --no-deps, --no-cache-dir, ...
+        pip_install_cmd = [
+            python,
+            "-m",
+            "pip",
+            "install",
+            "--disable-pip-version-check",
+        ] + pip_packages
+        logger.info("Installing python requirements to %s", virtualenv_path)
+        exit_code, output = exec_cmd_stream_to_logger(
+            pip_install_cmd, logger, cwd=cwd, env={})
+        if exit_code != 0:
+            raise RuntimeError(
+                f"Failed to install python requirements to {virtualenv_path}:\n{output}"
+            )
+
+    def run(self):
+        logger = self._logger
+        pip_packages = self._runtime_env.pip_packages()
+        path = _get_path_from_hash(self._pip_resources_dir,
+                                   _get_pip_hash(pip_packages))
+        # We create an empty directory for exec cmd so that the cmd will
+        # run more stable. e.g. if cwd has ray, then checking ray will
+        # look up ray in cwd instead of site packages.
+        exec_cwd = os.path.join(path, "exec_cwd")
+        os.makedirs(exec_cwd, exist_ok=True)
+        try:
+            virtualenv_path = self._create_or_get_virtualenv(
+                path, exec_cwd, logger)
+            python = self._get_virtualenv_python(virtualenv_path)
+            with self._check_ray(python, exec_cwd, logger):
+                self._install_pip_packages(virtualenv_path, pip_packages,
+                                           exec_cwd, logger)
+            # TODO(fyrestone): pip check.
+            self._context.py_executable = python
+        except Exception:
+            logger.info("Delete incomplete virtualenv: %s", path)
+            shutil.rmtree(path, ignore_errors=True)
+            logger.exception("Failed to install pip packages.")
+            raise
+
+
+class PipManager:
+    def __init__(self, resources_dir: str):
+        self._pip_resources_dir = os.path.join(resources_dir, "pip")
+        try_to_create_directory(self._pip_resources_dir)
 
     def delete_uri(self,
                    uri: str,
                    logger: Optional[logging.Logger] = default_logger) -> bool:
-        logger.info(f"Got request to delete URI {uri}")
+        logger.info("Got request to delete URI %s", uri)
         protocol, hash = parse_uri(uri)
         if protocol != Protocol.PIP:
             raise ValueError("PipManager can only delete URIs with protocol "
                              f"pip. Received protocol {protocol}, URI {uri}")
 
-        pip_env_path = self._get_path_from_hash(hash)
+        pip_env_path = _get_path_from_hash(self._pip_resources_dir, hash)
         try:
             shutil.rmtree(pip_env_path)
             successful = True
         except OSError:
             successful = False
-            logger.warning(f"Error when deleting pip env {pip_env_path}.")
+            logger.warning("Error when deleting pip env %s.", pip_env_path)
```

Comment:
Why not use the `f`-string?

Reply:
logger filters the log level first, then eval the var. Use % formatter can reduce the eval cost when the logging level mismatch. If we use f-string here, the `{pip_env_path}` will always be evaluated even we set the logging level to ERROR.

Reply:
Ah nice, I didn't know that!